File: android/app/src/main/java/com/voicecall/ui/screens/NotificationsScreen.kt
package com.voicecall.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.voicecall.R
import com.voicecall.data.Call
import com.voicecall.data.CallStatus
import com.voicecall.ui.CallsViewModel
import com.voicecall.ui.FeedRow
import com.voicecall.ui.GroupLabel
import com.voicecall.ui.GroupedCard
import com.voicecall.ui.ScreenHeader
import com.voicecall.ui.clickableNoRipple
import com.voicecall.ui.dayLabel
import com.voicecall.ui.formatClock
import com.voicecall.ui.navBarPadding
import com.voicecall.ui.statusBarPadding
import com.voicecall.ui.theme.Ink
import com.voicecall.ui.theme.Type
import com.voicecall.ui.theme.Wise

/**
 * What the assistant wants you to know.
 *
 * Derived from the calls rather than stored: every notification the prototype
 * shows — started, finished, nobody answered, needs you — is a transition this
 * app can already see in the call list, so there is nothing here the server has
 * to remember on our behalf.
 *
 * The one thing that genuinely needs storage is read state, and rather than fake
 * it this screen leaves it out: the hand-over banner stays at the top for as
 * long as the call is actually waiting, which is the only "unread" that matters.
 */
@Composable
fun NotificationsScreen(
    vm: CallsViewModel,
    onBack: () -> Unit,
    onOpenCall: (String) -> Unit,
    onOpenDetail: (String) -> Unit,
) {
    val state by vm.uiState.collectAsState()
    val context = LocalContext.current
    val now = remember { System.currentTimeMillis() }

    LaunchedEffect(Unit) { vm.refresh(quiet = true) }

    val waiting = state.calls.filter { it.status == CallStatus.TRANSFERRING }
    val rest = state.calls.filterNot { it.status == CallStatus.TRANSFERRING }

    Column(
        Modifier
            .fillMaxSize()
            .padding(top = statusBarPadding(), bottom = navBarPadding()),
    ) {
        ScreenHeader(title = stringResource(R.string.nav_notifications), onBack = onBack)

        Column(
            Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .padding(start = 20.dp, end = 20.dp, top = 14.dp, bottom = 24.dp),
        ) {
            // A call holding the line for you is not a list item. It is the
            // reason the app has a notification screen at all.
            waiting.forEach { call ->
                HandoverBanner(call) { onOpenCall(call.id) }
                Spacer(Modifier.height(12.dp))
            }

            if (rest.isEmpty() && waiting.isEmpty()) {
                Text(stringResource(R.string.notifs_empty), style = Type.Caption, color = Ink.Mute)
            }

            rest.groupBy { dayLabel(context, it.createdAt, now) }.forEach { (day, calls) ->
                GroupLabel(day, modifier = Modifier.padding(top = 6.dp, bottom = 8.dp))
                GroupedCard(rows = calls.map { call ->
                    {
                        FeedRow(
                            dot = dotFor(call),
                            title = headlineFor(call),
                            subtitle = call.summary?.ifBlank { null } ?: call.goal,
                            pulsing = call.isLive,
                            onClick = { if (call.isLive) onOpenCall(call.id) else onOpenDetail(call.id) },
                            right = {
                                Text(
                                    formatClock(call.endedAt ?: call.createdAt),
                                    style = Type.Mono,
                                    color = Ink.Rim,
                                )
                            },
                        )
                    }
                })
                Spacer(Modifier.height(16.dp))
            }
        }
    }
}

/** Amber, loud, and the only thing on the screen with a button in it. */
@Composable
private fun HandoverBanner(call: Call, onOpen: () -> Unit) {
    Column(
        Modifier
            .fillMaxWidth()
            .background(Ink.Warning, RoundedCornerShape(16.dp))
            .padding(16.dp),
    ) {
        Row {
            Box(
                Modifier.padding(top = 6.dp).size(8.dp).background(Ink.WarningDeep, CircleShape),
            )
            Spacer(Modifier.width(12.dp))
            Column(Modifier.weight(1f)) {
                Text(
                    stringResource(R.string.headline_needs_you, call.businessName),
                    style = Type.RowTitle,
                    color = Ink.WarningInk,
                )
                Spacer(Modifier.height(3.dp))
                Text(
                    stringResource(R.string.subline_transfer),
                    style = Type.RowSub,
                    color = Ink.WarningInk,
                )
                Spacer(Modifier.height(10.dp))
                Row(
                    Modifier
                        .border(1.5.dp, Ink.WarningInk, RoundedCornerShape(10.dp))
                        .clickableNoRipple(onOpen)
                        .padding(horizontal = 14.dp, vertical = 8.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(Wise.Phone, null, tint = Ink.WarningInk, modifier = Modifier.size(13.dp))
                    Spacer(Modifier.width(7.dp))
                    Text(
                        stringResource(R.string.action_bridge_me),
                        style = Type.ChipStrong,
                        color = Ink.WarningInk,
                    )
                }
            }
        }
    }
}

private fun dotFor(call: Call): Color = when {
    call.isLive -> Ink.Lime
    call.status == CallStatus.FAILED -> Ink.Negative
    call.outcome == "failed" -> Ink.Negative
    call.status == CallStatus.COMPLETED -> Ink.Positive
    else -> Ink.Rim
}

@Composable
private fun headlineFor(call: Call): String = when {
    call.isLive -> stringResource(R.string.notifs_started, call.businessName)
    call.status == CallStatus.FAILED -> stringResource(R.string.notifs_no_answer, call.businessName)
    else -> stringResource(R.string.notifs_finished, call.businessName)
}

File: ios/Mynah/UI/Screens/NotificationsScreen.swift
import SwiftUI

/// What the assistant wants you to know.
///
/// Derived from the calls rather than stored: every notification the prototype
/// shows — started, finished, nobody answered, needs you — is a transition this
/// app can already see in the call list, so there is nothing here the server has
/// to remember on our behalf.
///
/// The one thing that genuinely needs storage is read state, and rather than
/// fake it this screen leaves it out: the hand-over banner stays at the top for
/// as long as the call is actually waiting, which is the only "unread" that
/// matters.
struct NotificationsScreen: View {
    @ObservedObject var model: CallsViewModel
    let onBack: () -> Void
    let onOpenCall: (String) -> Void
    let onOpenDetail: (String) -> Void

    @State private var now = Int(Date().timeIntervalSince1970 * 1000)

    private var waiting: [Call] { model.calls.filter { $0.status == CallStatus.transferring } }
    private var rest: [Call] { model.calls.filter { $0.status != CallStatus.transferring } }

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(t("nav_notifications"), onBack: onBack)

            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    // A call holding the line for you is not a list item. It is
                    // the reason the app has a notification screen at all.
                    ForEach(waiting) { call in
                        HandoverBanner(call: call) { onOpenCall(call.id) }
                        Spacer().frame(height: 12)
                    }

                    if rest.isEmpty, waiting.isEmpty {
                        Text(t("notifs_empty")).wise(Type.caption).foregroundStyle(Ink.mute)
                    }

                    ForEach(days, id: \.day) { group in
                        GroupLabel(group.day)
                            .padding(.top, 6)
                            .padding(.bottom, 8)
                        GroupedCard {
                            ForEach(group.calls) { call in
                                FeedRow(
                                    dot: dot(for: call),
                                    title: headline(for: call),
                                    subtitle: call.summary?.nilIfBlank ?? call.goal,
                                    onTap: { call.isLive ? onOpenCall(call.id) : onOpenDetail(call.id) },
                                    pulsing: call.isLive,
                                    right: {
                                        Text(formatClock(call.endedAt ?? call.createdAt))
                                            .wise(Type.mono)
                                            .foregroundStyle(Ink.rim)
                                    }
                                )
                            }
                        }
                        Spacer().frame(height: 16)
                    }
                }
                .padding(.horizontal, 20)
                .padding(.top, 14)
                .padding(.bottom, 24)
            }
        }
        .navigationBarBackButtonHidden()
        .onAppear { model.refresh(quiet: true) }
    }

    private var days: [(day: String, calls: [Call])] {
        var order: [String] = []
        var grouped: [String: [Call]] = [:]
        for call in rest {
            let day = dayLabel(call.createdAt, now: now, language: model.language)
            if grouped[day] == nil { order.append(day) }
            grouped[day, default: []].append(call)
        }
        return order.map { ($0, grouped[$0] ?? []) }
    }

    private func dot(for call: Call) -> Color {
        if call.isLive { return Ink.lime }
        if call.status == CallStatus.failed || call.outcome == "failed" { return Ink.negative }
        if call.status == CallStatus.completed { return Ink.positive }
        return Ink.rim
    }

    private func headline(for call: Call) -> String {
        if call.isLive { return t("notifs_started", call.businessName) }
        if call.status == CallStatus.failed { return t("notifs_no_answer", call.businessName) }
        return t("notifs_finished", call.businessName)
    }
}

/// Amber, loud, and the only thing on the screen with a button in it.
private struct HandoverBanner: View {
    let call: Call
    let onOpen: () -> Void

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Circle()
                .fill(Ink.warningDeep)
                .frame(width: 8, height: 8)
                .padding(.top, 6)
            VStack(alignment: .leading, spacing: 0) {
                Text(t("headline_needs_you", call.businessName))
                    .wise(Type.rowTitle)
                    .foregroundStyle(Ink.warningInk)
                    .fixedSize(horizontal: false, vertical: true)
                Spacer().frame(height: 3)
                Text(t("subline_transfer"))
                    .wise(Type.rowSub)
                    .foregroundStyle(Ink.warningInk)
                    .fixedSize(horizontal: false, vertical: true)
                Spacer().frame(height: 10)
                HStack(spacing: 7) {
                    Icon(Wise.phone, size: 13).foregroundStyle(Ink.warningInk)
                    Text(t("action_bridge_me")).wise(Type.chipStrong).foregroundStyle(Ink.warningInk)
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 8)
                .overlay(
                    RoundedRectangle(cornerRadius: 10, style: .continuous)
                        .strokeBorder(Ink.warningInk, lineWidth: 1.5)
                )
                .contentShape(Rectangle())
                .onTapGesture(perform: onOpen)
            }
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Ink.warning, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
    }
}
